Display site name screen in site creation flow

mkevins opened the pull request:
Fixes #16245, #16246, #16248

### Description

This PR adds the site name screen (behind a feature flag and ExPlat experiment).



**Note**:

* This currently resolves the feature flag and ExPlat tasks as well. If it helps to review in smaller pieces, I've broken the commits (https://github.com/wordpress-mobile/WordPress-Android/pull/16259/commits/48f9ce85959cfe4c9e5065d2678815f4fa693483, https://github.com/wordpress-mobile/WordPress-Android/pull/16259/commits/89e857aaa2b8de17e1a5e838c62d5b0f3be54d8e) up so they can be merged in separate PRs if needed.
* This commit (https://github.com/wordpress-mobile/WordPress-Android/pull/16259/commits/a793e89a79c92e381358a4c9ac1aef167d0b78a6) can be reverted to hard-code the flag to `true` for convenience while testing.
  * Also, the behavior remains conditioned behind the site intent question flag
* The experiment name is: `"wpandroid_site_name_experiment"`, we can change this if necessary

**Not yet implemented**:
* <s>The continue button is enabled when the input is empty</s>
  * This should be disabled (or hidden?) when the input is empty
  * I have hidden the continue button, but we can change to a disabled state (and style) if needed
* Tracks events are not implemented (will be addressed in a future PR)
  * I left a question about whether we'd want to be tracking site names entered, or just the event itself
  * In places were tracking would occur, I placed `Log` statements instead to aid in manual testing

### Testing

#### Feature flag & ExPlat

Note: It may be helpful to manually hard-code a value for `true` for the ExPlat flag if testing before that is established server-side.

* Toggle the feature flag: Me -> App settings -> Debug settings -> Toggle flag (`SiteNameFeatureConfig`)
* Navigate through the site creation flow: My Sites -> `+` -> Create WordPress.com site -> Choose an intent
* Expect the site name screen to appear after the intent screen
* Toggle the feature off
* Expect the site designs screen to appear after the intent screen

#### Site Screen UI

Note: for these tests, the stubs for tracking can be tested by filtering Logcat for `siteNameWip`

##### Continuing

1. Perform the steps above to navigate to the site name screen
  * The logs should show that the site name screen has been viewed
  * The input should be focused and soft keyboard opened (if a physical keyboard is not present)
  * The Continue button should not be visible
2. Type in a site name
  * The Continue button should become visible
3. Tap Continue
  * The logs should show that the site name was entered
  * The site design screen should be shown

##### Skipping

1. Perform the steps above to navigate to the site name screen
  * The logs should show that the site name screen has been viewed
  * The input should be focused and soft keyboard opened (if a physical keyboard is not present)
  * The Continue button should not be visible
2. Tap Skip
  * The logs should show that the site name was skipped
  * The site design screen should be shown

##### Backtracking

1. Perform the steps above to navigate to the site name screen
  * The logs should show that the site name screen has been viewed
  * The input should be focused and soft keyboard opened (if a physical keyboard is not present)
  * The Continue button should not be visible
2. Type in a site name
  * The Continue button should become visible
3. Tap the back button
  *  The site intent screen should be shown
4. Choose another intent (could be the same one, it doesn't matter)
  * The site name screen should be shown with the previously entered site name (i.e. backtracking should not delete previously entered form data)

### Screenshots

| Empty Input | Filled Input |
|-|-|
|![site-name-screen-empty](https://user-images.githubusercontent.com/8507675/161929278-b1aec40c-15ab-415b-aa04-bfa287471594.png)|![site-name-screen-filled](https://user-images.githubusercontent.com/8507675/161929296-b6ade85f-400f-422e-b5e5-89402fe54fc5.png)|



## Regression Notes
1. Potential unintended areas of impact
This could affect the site creation flow

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Existing automated tests for the site creation flow provide good coverage

3. What automated tests I added (or what prevented me from doing so)
More coverage for these changes to the flow will come in later PRs (this is currently behind a feature flag)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

mkevins committed:
Add site name feature flag

mkevins committed:
Add site name ExPlat experiment

mkevins committed:
Add site name view model

mkevins committed:
Add site name fragment

mkevins committed:
Rename SEGMENTS to SITE_DESIGNS

mkevins committed:
Add site name step and wire up to site creation

ovitrif commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/sitename/SiteCreationSiteNameFragment.kt:
```diff
@@ -0,0 +1,97 @@
+        binding = null
+    }
+
+    private fun isPhoneLandscape() = displayUtils.isLandscapeBySize() && !displayUtils.isTablet()
```
Seeing that `isPhoneLandscape()` is quite a "popular" need I was thinking maybe we can extract it to `displayUtils`.

Wdyt?

mkevins replied:
Good point Ovi. Not sure whether or not we'll end up needing this, but if we keep it, I will extract it to `displayUtils` :+1:

mkevins committed:
Explicitly set nulls when skipping

This is important in case the user backtracks and then decides to skip,
since the final site creation data should reflect that intention.
Without this, we'd still have the stale data at the end of the flow.
While only using this for tracking so far, we have not encountered this,
but this staleness would be an issue in the future.

mkevins committed:
Hide continue button when input is empty

This currently hides the input button. Alternatively, we can also
disable it and provide styles for that.

mkevins committed:
Manage keyboard and focus between screens

This auto-focuses the input, since it is the only input on the screen.
It also manages the visibility of the keyboard, and restores the input
text contents from the view model (important when a user back tracks to
a previous step in the flow and then wants to continue forward).

mkevins commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationActivity.kt:
```diff
@@ -131,6 +146,11 @@ class SiteCreationActivity : LocaleAwareActivity(),
         ActivityUtils.hideKeyboard(this)
```
I'm wondering if we should use the feature / ExPlat flag(s) here to prevent keyboard bounce. :thinking: It is a minor issue:

* When the site name is OFF, site designs follow intents (so it makes sense to hide)
* When the site name is ON, site name follows intents (so we probably want to keep the keyboard visible)

antonis replied:
I think it's a good idea to prevent the bounce 👍

mkevins replied:
:+1: Added here: https://github.com/wordpress-mobile/WordPress-Android/pull/16259/commits/5f66466996890712c37cb22c0ea677256bc3cb47

mkevins committed:
Remove hard-coded flag values

This commit can be reverted for ease of testing the UI parts of the PR.

antonis commented on WordPress/src/main/java/org/wordpress/android/util/experiments/SiteNameABExperiment.kt:
```diff
@@ -0,0 +1,6 @@
+import javax.inject.Inject
+
+class SiteNameABExperiment
+@Inject constructor(exPlat: ExPlat) : Experiment(name = "wpandroid_site_name_experiment", exPlat)
```
Created an experiment named `wpandroid_site_name_v1` before I got a look at the name here :(

antonis replied:
Updated with 8f361d4672f53abcd96c4f02d82c23d2b59c7d28

mkevins replied:
Thanks for updating (and also fixing the ktlint :+1: )

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/sitecreation/SiteCreationStep.kt:
```diff
@@ -1,34 +1,49 @@
 ) {
     fun getSteps(): List<SiteCreationStep> {
         if (siteIntentQuestionFeatureConfig.isEnabled()) {
+             if (siteNameFeatureConfig.isEnabled() && siteNameABExperiment.getVariation() != Control) {
```
I wonder if we can keep the `siteIntentQuestionFeatureConfig` independent of the `siteNameFeatureConfig`. This is not blocking for this PR but maybe it will be more versatile if we composed the list more dynamically (e.g. add items to a mutable list when a dynamic feature is available)

antonis replied:
Since `Detekt` also complains about `ReturnCount` I was thinking of something like
```
    fun getSteps() = mutableListOf<SiteCreationStep>().apply {
        if (siteIntentQuestionFeatureConfig.isEnabled()) {
            add(INTENTS)
        }
        if (siteNameFeatureConfig.isEnabled() && siteNameABExperiment.getVariation() != Control) {
            add(SITE_NAME)
        }
        add(SITE_DESIGNS)
        add(DOMAINS)
        add(SITE_PREVIEW)
    }.toList()
```
Wdyt?

mkevins replied:
I made this dependent in anticipation of the heading enhancement (make it dependent on the intent screen). But the flag itself doesn't need to be dependent, instead we can include the `INTENTS` with the site name flow variation. I've added this here: https://github.com/wordpress-mobile/WordPress-Android/pull/16259/commits/35edfb6a6057e8c323b14cd1a988c0d4a2ba7cc5 as a when expression, as an alternative (since this makes it nice to compare the lists, imo). Wdyt?

antonis replied:
Thank you for fixing this @mkevins. The solution with the `when` clause makes it even more clear 👍

antonis committed:
Changed experiment name

antonis committed:
Removes trailing comma not supported by ktlint

antonis reviewed (commented):
Hey @mkevins 👋 
great work on this. I've left a few comments. The only thing blocking merge is the [Detekt issue](https://github.com/wordpress-mobile/WordPress-Android/pull/16259/files#r844141312) for which I added a suggestion.

I'd also like @kyleaparker's feedback on the Continue button visibility. This deviates from the design (ref `pbArwn-447-p2`) that has a disabled button state instead of having it invisible. It aligns though with the current behaviour of the domain screen. Personally I'm ok with both approaches 🤔 

<details>
<summary>Continue button visibility recording</summary>

https://user-images.githubusercontent.com/304044/162022156-0b7b070f-02e1-439e-8515-460af82e3c39.mp4
<details>

mkevins committed:
Reduce returns to satisfy detekt

mkevins committed:
Fix lint issue

mkevins committed:
Fix keyboard bounce when site name screen follows intents screen

antonis reviewed (approved):
Awesome work @mkevins 👍 
I retested with the latest commits and everything LGTM 🎉

kyleaparker commented:
> I'd also like @kyleaparker's feedback on the Continue button visibility. This deviates from the design (ref pbArwn-447-p2) that has a disabled button state instead of having it invisible. It aligns though with the current behaviour of the domain screen. Personally I'm ok with both approaches 🤔

I'm OK with this approach 👍
